provider/plugin/httpapi_test.go

```diff
@@ -0,0 +1,200 @@
+	changes := &plan.Changes{
+		Create: []*endpoint.Endpoint{
+			{
+				DNSName:    "foo.bar.com",
+				RecordType: "A",
+				Targets:    endpoint.Targets{},
+			},
+		},
+	}
+	j, err := json.Marshal(changes)
+	require.Nil(t, err)
+
+	reader := bytes.NewReader(j)
+
+	req := httptest.NewRequest(http.MethodPost, "/applychanges", reader)
+	w := httptest.NewRecorder()
+
+	httpProvider := &HTTPProvider{
+		provider: &FakePluginProvider{},
+	}
+	httpProvider.recordsHandler(w, req)
+	res := w.Result()
+	require.Equal(t, http.StatusOK, res.StatusCode)
+}
+
+func TestPropertyValuesEqualHandlerWithInvalidRequests(t *testing.T) {
+	req := httptest.NewRequest(http.MethodPost, "/propertyvaluesequals", nil)
+	w := httptest.NewRecorder()
+
+	httpProvider := &HTTPProvider{
+		provider: &FakePluginProvider{},
+	}
+	httpProvider.propertyValuesEqualHandler(w, req)
+	res := w.Result()
+	require.Equal(t, http.StatusBadRequest, res.StatusCode)
+
+	req = httptest.NewRequest(http.MethodGet, "/propertyvaluesequals", nil)
+
+	httpProvider.propertyValuesEqualHandler(w, req)
+	res = w.Result()
+	require.Equal(t, http.StatusBadRequest, res.StatusCode)
+}
+
+func TestPropertyValuesEqualWithValidRequest(t *testing.T) {
+	pve := &PropertyValuesEqualsRequest{
+		Name:     "foo",
+		Previous: "bar",
+		Current:  "baz",
+	}
+
+	j, err := json.Marshal(pve)
+	require.Nil(t, err)
+
+	reader := bytes.NewReader(j)
+	req := httptest.NewRequest(http.MethodPost, "/propertyvaluesequals", reader)
+	w := httptest.NewRecorder()
+
+	httpProvider := &HTTPProvider{
+		provider: &FakePluginProvider{},
+	}
+	httpProvider.propertyValuesEqualHandler(w, req)
+	res := w.Result()
+	require.Equal(t, http.StatusOK, res.StatusCode)
+	require.NotNil(t, res.Body)
+}
+
+func TestAdjustEndpointsHandlerWithInvalidRequest(t *testing.T) {
+	req := httptest.NewRequest(http.MethodPost, "/adjustendpoints", nil)
+	w := httptest.NewRecorder()
+
+	httpProvider := &HTTPProvider{
+		provider: &FakePluginProvider{},
+	}
+	httpProvider.adjustEndpointsHandler(w, req)
+	res := w.Result()
+	require.Equal(t, http.StatusBadRequest, res.StatusCode)
+
+	req = httptest.NewRequest(http.MethodGet, "/adjustendpoints", nil)
+
+	httpProvider.adjustEndpointsHandler(w, req)
+	res = w.Result()
+	require.Equal(t, http.StatusBadRequest, res.StatusCode)
+}
+
+func TestAdjustEndpointsWithValidRequest(t *testing.T) {
+	pve := []*endpoint.Endpoint{
+		{
+			DNSName:    "foo.bar.com",
+			RecordType: "A",
+			Targets:    endpoint.Targets{},
+			RecordTTL:  0,
+		},
+	}
+
+	j, err := json.Marshal(pve)
+	require.Nil(t, err)
+
+	reader := bytes.NewReader(j)
+	req := httptest.NewRequest(http.MethodPost, "/adjustendpoints", reader)
+	w := httptest.NewRecorder()
+
+	httpProvider := &HTTPProvider{
+		provider: &FakePluginProvider{},
+	}
+	httpProvider.adjustEndpointsHandler(w, req)
+	res := w.Result()
+	require.Equal(t, http.StatusOK, res.StatusCode)
+	require.NotNil(t, res.Body)
+}
+
+func TestNegotiate(t *testing.T) {
```

Comment:
Maybe also add a test what happens on wrong version by the client, as this was a comment already we should make sure it's defined in some way.

Reply:
makes sense.